after upgrading 7.8.13 to 7.8.22 getting Invalid File Reference from attachments in inbound

might be related to #6418
#### Issue
inbound email do not store attachments properly, they log in database but do not store in filesystem or upload directory.
running IIs should not be a security issue because if I write an email with attachment the outgoing attachment is recorded properly in the notes table and attachment is found in upload directory.


#### Expected Behavior
store attachments in upload directory as the notes record shows, currently the received attachment appear to be lost.

#### Actual Behavior
notes table is updated but no file present

#### Possible Fix
<!--- Not obligatory, but suggest a fix or reason for the bug -->

#### Steps to Reproduce
send and email to an inbound address that creates cases and try to look at the attachment in crm...
Invalid File Reference

high priority
#### Context
we can no longer process orders received by inbound emails

#### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
 SuiteCRM Version used: 7.8.22
not related to browser  :
MSsql server 
IIs php 7.1

We have also detected the same problem after 7.8.20->7.8.24 upgrade. The following is observed: 
- Inbound imails arrive correctly in CRM personal inbox. 
- Emails can be viewed and email text displays correctly. When new email via CRM emails, the folloing Error messages are observed in crm-log file.  (CRM log level is set to Warn). 
`Thu Jan 17 10:21:32 2019 [4952][skip][WARN] InboundEmail::connectMailserver: Request ssl is not defined
Thu Jan 17 10:21:32 2019 [4952][skip][WARN] Saving error level. Try to remove the error_reporting() function from your code.
Thu Jan 17 10:21:32 2019 [4952][skip][ERROR] Pop error level. Try to remove the error_reporting() function from your code.
Thu Jan 17 10:21:32 2019 [4952][skip][WARN] InboundEmail::handleEncodedFilename: unable to resolve encoding
Thu Jan 17 10:21:32 2019 [4952][skip][WARN] InboundEmail::saveAttachmentBinaries: file not found: cache/modules/Emails/skip-c8b7-60f5-skip-56331d12d8ca/attachments/26aee1a7432306df975db12e7b6848d60
Thu Jan 17 10:21:32 2019 [4952][skip][WARN] InboundEmail::saveAttachmentBinaries: file not found: cache/modules/Emails/skip-c8b7-60f5-skip-56331d12d8ca/attachments/26aee1a7432306df975db12e7b6848d61
Thu Jan 17 10:21:32 2019 [4952][skip][WARN] InboundEmail::handleEncodedFilename: unable to resolve encoding
Thu Jan 17 10:21:32 2019 [4952][skip][WARN] InboundEmail::saveAttachmentBinaries: file not found: cache/modules/Emails/skip-c8b7-60f5-skip-56331d12d8ca/attachments/26aee1a7432306df975db12e7b6848d62
Thu Jan 17 10:21:32 2019 [4952][skip][WARN] InboundEmail::saveAttachmentBinaries: file not found: cache/modules/Emails/skip-c8b7-60f5-skip-56331d12d8ca/attachments/26aee1a7432306df975db12e7b6848d63
Thu Jan 17 10:21:32 2019 [4952][skip][WARN] InboundEmail::handleEncodedFilename: unable to resolve encoding
Thu Jan 17 10:21:33 2019 [4952][skip][WARN] InboundEmail::saveAttachmentBinaries: file not found: cache/modules/Emails/skip-c8b7-60f5-skip-56331d12d8ca/attachments/26aee1a7432306df975db12e7b6848d64
Thu Jan 17 10:21:33 2019 [4952][skip][WARN] InboundEmail::saveAttachmentBinaries: file not found: cache/modules/Emails/skip-c8b7-60f5-skip-56331d12d8ca/attachments/26aee1a7432306df975db12e7b6848d65
Thu Jan 17 10:21:33 2019 [4952][skip][WARN] InboundEmail::handleEncodedFilename: unable to resolve encoding
Thu Jan 17 10:21:33 2019 [4952][skip][WARN] InboundEmail::saveAttachmentBinaries: file not found: cache/modules/Emails/skip-c8b7-60f5-skip-56331d12d8ca/attachments/26aee1a7432306df975db12e7b6848d66
Thu Jan 17 10:21:33 2019 [4952][skip][WARN] InboundEmail::saveAttachmentBinaries: file not found: cache/modules/Emails/skip-c8b7-60f5-skip-56331d12d8ca/attachments/26aee1a7432306df975db12e7b6848d67
Thu Jan 17 10:21:33 2019 [4952][skip][WARN] InboundEmail::connectMailserver: Request ssl is not defined
Thu Jan 17 10:21:33 2019 [4952][skip][WARN] Saving error level. Try to remove the error_reporting() function from your code.
Thu Jan 17 10:21:33 2019 [4952][skip][ERROR] Pop error level. Try to remove the error_reporting() function from your code.`
- When clicking on email attachment (from within the CRM emails) a new blank page opens displaying text "Invalid File Reference".  Nothing is logged in the crm-log file. 
- When attaching email to a CRM record (e.g. Importing email to account (from CRM emails)),  email record attaches to the account. The following is logged to crm error log file: 
 `
Thu Jan 17 10:24:45 2019 [4919][skip][WARN] InboundEmail::connectMailserver: Request ssl is not defined
Thu Jan 17 10:24:45 2019 [4919][skip][WARN] Saving error level. Try to remove the error_reporting() function from your code.
Thu Jan 17 10:24:46 2019 [4919][skip][ERROR] Pop error level. Try to remove the error_reporting() function from your code.
Thu Jan 17 10:24:46 2019 [4919][skip][WARN] InboundEmail::handleEncodedFilename: unable to resolve encoding
Thu Jan 17 10:24:46 2019 [4919][skip][WARN] InboundEmail::saveAttachmentBinaries: file not found: upload://28f4baa9-8aee-d6fe-083a-ae4057620475
Thu Jan 17 10:24:46 2019 [4919][skip][WARN] InboundEmail::saveAttachmentBinaries: file not found: upload://3a56c0ab-d421-c3f6-40c1-5c405eab77dd
Thu Jan 17 10:24:46 2019 [4919][skip][WARN] InboundEmail::handleEncodedFilename: unable to resolve encoding
Thu Jan 17 10:24:46 2019 [4919][skip][WARN] InboundEmail::saveAttachmentBinaries: file not found: upload://42c9cfcb-a1ee-663f-5b50-5c40571af6df
Thu Jan 17 10:24:46 2019 [4919][skip][WARN] InboundEmail::saveAttachmentBinaries: file not found: upload://56ec8e78-8ee6-3e67-7a00-5c4057aa5b1b
Thu Jan 17 10:24:46 2019 [4919][skip][WARN] InboundEmail::handleEncodedFilename: unable to resolve encoding
Thu Jan 17 10:24:46 2019 [4919][skip][WARN] InboundEmail::saveAttachmentBinaries: file not found: upload://5f82a62b-eds2-89ce-c25c-5c40574e33c0
Thu Jan 17 10:24:46 2019 [4919][skip][WARN] InboundEmail::saveAttachmentBinaries: file not found: upload://682b90fe-3120-26bb-9d43-5c40574915ab
Thu Jan 17 10:24:46 2019 [4919][skip][WARN] InboundEmail::handleEncodedFilename: unable to resolve encoding
Thu Jan 17 10:24:46 2019 [4919][skip][WARN] InboundEmail::saveAttachmentBinaries: file not found: upload://70c6f86d-732e-c213-1427-5c4057ae9876
Thu Jan 17 10:24:46 2019 [4919][skip][WARN] InboundEmail::saveAttachmentBinaries: file not found: upload://791ac730-889e-124f-7d13-5c4057ea5e51
Thu Jan 17 10:24:46 2019 [4919][skip][DEPRECATED] Formatting correction: Emails->date_sent had formatting automatically corrected. This will be removed in the future, please upgrade your external code
Thu Jan 17 10:24:46 2019 [4919][skip][WARN] CaseUpdatesHook: saveEmailUpdate: Not a create case or wrong parent type`


We have a CRM instance available where we could assist with testing a bug-fix.  This bug is of somewhat high priority to us as it is impacting client data integrity for received emails.   

SuiteCRM 7.8.24
MySQLd 5.6.42
PHP 7.0.33
CentOS Linux 7.6.1810

@Dillon-Brown, do you have an indication when this defect could tiaged for development? Thanks in advance!!

--- a/modules/InboundEmail/InboundEmail.php
+++ b/modules/InboundEmail/InboundEmail.php
@@ -3777,78 +3777,61 @@ class InboundEmail extends SugarBean {
 	   return $result;
 
     }
-	/**
-	 * saves the actual binary file of a given attachment
-	 * @param object attach Note object that is attached to the binary file
-	 * @param string msgNo Message Number on IMAP/POP3 server
-	 * @param string thisBc Breadcrumb to navigate email structure to find the content
-	 * @param object part IMAP standard object that contains the "parts" of this section of email
-	 * @param bool $forDisplay
-	 */
-	function saveAttachmentBinaries($attach, $msgNo, $thisBc, $part, $forDisplay) {
-		// decide where to place the file temporarily
 
-		if(isset($attach->id) && strpos($attach->id, "..") !== false && isset($this->id) && strpos($this->id, "..") !== false){
-			die("Directory navigation attack denied.");
-		}
+    /**
+     * @param $attach object Note object that is attached to the binary file
+     * @param $msgNo string Message Number on IMAP/POP3 server
+     * @param $thisBc string Breadcrumb to navigate email structure to find the content
+     * @param $part object IMAP standard object that contains the "parts" of this section of email
+     * @param $forDisplay
+     * @return bool|void
+     */
+    public function saveAttachmentBinaries($attach, $msgNo, $thisBc, $part, $forDisplay)
+    {
+        if (isset($attach->id) &&
+            strpos($attach->id, '..') !== false &&
+            isset($this->id) &&
+            strpos($this->id, '..') !== false
+        ) {
+            die('Directory navigation attack denied.');
+        }
+        $uploadDir = $forDisplay ? "{$this->EmailCachePath}/{$this->id}/attachments/" : "upload://";
+        $fileName = htmlspecialchars($attach->id);
+        if (!file_exists($uploadDir . $fileName)) {
+            if (!is_resource($this->conn)) {
+                LoggerManager::getLogger()->fatal('Inbound Email Connection is not valid resource for saving attachment binaries.');
 
-		$uploadDir = ($forDisplay) ? "{$this->EmailCachePath}/{$this->id}/attachments/" : "upload://";
+                return false;
+            }
 
-		// decide what name to save file as
-		$fileName = htmlspecialchars($attach->id);
+            if (null === $this->conn) {
+                LoggerManager::getLogger()->warn('InboundEmail::saveAttachmentBinaries: connection is null');
+                $msgPartRaw = null;
+            } else {
+                $msgPartRaw = imap_fetchbody($this->conn, $msgNo, $thisBc);
+            }
+            // deal with attachment encoding and decode the text string
+            $msgPart = $this->handleTranserEncoding($msgPartRaw, $part->encoding);
+            if (file_put_contents($uploadDir . $fileName, $msgPart)) {
+                $GLOBALS['log']->debug('InboundEmail saved attachment file: ' . $attach->filename);
+            } else {
+                $GLOBALS['log']->debug('InboundEmail could not create attachment file: ' . $attach->filename . " - temp file target: [ {$uploadDir}{$fileName} ]");
 
-		// download the attachment if we didn't do it yet
-		if(!file_exists($uploadDir.$fileName)) {
-                    
-                    if (null === $this->conn) {
-                        LoggerManager::getLogger()->warn('InboundEmail::saveAttachmentBinaries: connection is null');
-                        $msgPartRaw = null;
-                    } else {
-			$msgPartRaw = imap_fetchbody($this->conn, $msgNo, $thisBc);
-                    }
-                    
-    		// deal with attachment encoding and decode the text string
-                    
-                    $partEncoding = null;
-                    if (isset($part->encoding)) {
-                        $partEncoding = $part->encoding;
-                    } else {
-                        LoggerManager::getLogger()->warn('InboundEmail::saveAttachmentBinaries: part encoding is not set');
-                    }
-                    
-			$msgPart = $this->handleTranserEncoding($msgPartRaw, $partEncoding);
-
-                        $file = $uploadDir.$fileName;
-                        if (!file_exists($file)) {
-                            LoggerManager::getLogger()->warn('InboundEmail::saveAttachmentBinaries: file not found: ' . $file);
-                        } elseif(file_put_contents($file, $msgPart)) {
-				$GLOBALS['log']->debug('InboundEmail saved attachment file: '.$attach->filename);
-			} else {
-                $GLOBALS['log']->debug('InboundEmail could not create attachment file: '.$attach->filename ." - temp file target: [ {$uploadDir}{$fileName} ]");
                 return;
-			}
-		}
-
-		$this->tempAttachment[$fileName] = urldecode($attach->filename);
-		// if all was successful, feel for inline and cache Note ID for display:
-                
-                $partType = null;
-                if (!isset($part->type)) {
-                    LoggerManager::getLogger()->warn('InboundEmail::saveAttachmentBinaries: part type is not set');
-                } else {
-                    $partType = $part->type;
-                }
-                
-		if((strtolower($part->disposition) == 'inline' && in_array($part->subtype, $this->imageTypes))
-		    || ($partType == 5)) {
-		    if(copy($uploadDir.$fileName, sugar_cached("images/{$fileName}.").strtolower($part->subtype))) {
-			    $id = substr($part->id, 1, -1); //strip <> around
-			    $this->inlineImages[$id] = $attach->id.".".strtolower($part->subtype);
-			} else {
-				$GLOBALS['log']->debug('InboundEmail could not copy '.$uploadDir.$fileName.' to cache');
-			}
-		}
-	}
+            }
+        }
+        $this->tempAttachment[$fileName] = urldecode($attach->filename);
+        if ((strtolower($part->disposition) === 'inline' && in_array($part->subtype, $this->imageTypes, false))
+            || ($part->type == 5)
+        ) {
+            if (copy($uploadDir . $fileName, sugar_cached("images/{$fileName}.") . strtolower($part->subtype))) {
+                $id = substr($part->id, 1, -1); //strip <> around
+                $this->inlineImages[$id] = $attach->id . "." . strtolower($part->subtype);
+            } else {
+                $GLOBALS['log']->debug('InboundEmail could not copy ' . $uploadDir . $fileName . ' to cache');
+            }
+        }
+    }
 
 	/**
 	 * decodes a string based on its associated encoding

